File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-2.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        mp = {}
        l = 0
        res = 0

        for i in range(len(s)):
            if s[i] in mp and mp[s[i]] >= l:
                l = mp[s[i]] + 1
            mp[s[i]] = i
            res = max(res, i - l + 1)

        return res

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int res = INT_MIN;
    int maxPathSum(TreeNode* root) {
        return max(res, util(root));
    }

    int util(TreeNode* root) {
        int left = 0, right = 0;
        if(root->left)
        left = util(root->left);
        if(root->right)
        right = util(root->right);
        res = max(max(max(res, left+right+root->val), root->val), max(left, right) + root->val);
        return max((max(left, right) + root->val), root->val);
    }
};
